### PR TITLE
Support einsum optimize by greedy method.

### PR DESCRIPTION
**Pull request by fukatani**
Support numpy like einsum optimization.

See https://github.com/dgasmith/opt_einsum
I will implement 'greedy' option only in this PR.

### CODE REVIEW

**Comment by fukatani** [comment]
I confirmed optimization effect by following script.

```python
N = 4
C = cupy.random.rand(N, N)
I = cupy.random.rand(N, N, N, N)

cnt = 100

def naive(I, C):
    return cupy.einsum('pi,qj,ijkl,rk,sl->pqrs', C, C, I, C, C)

def optimize(I, C):
    return cupy.einsum('pi,qj,ijkl,rk,sl->pqrs', C, C, I, C, C, optimize=True)

stream = cupy.cuda.Stream.null
start = stream.record()
for i in range(cnt):
    naive(I, C)
end = stream.record()
end.synchronize()
print('Naive cupy einsum', cupy.cuda.get_elapsed_time(start, end) / cnt)

stream = cupy.cuda.Stream.null
start = stream.record()
for i in range(cnt):
    optimize(I, C)
end = stream.record()
end.synchronize()
print('Optimized cupy einsum', cupy.cuda.get_elapsed_time(start, end) / cnt)
```

$ Naive cupy einsum 14.9975146484375 (ms)
$ Optimized cupy einsum 3.9200555419921876 (ms)

Memory consumption is also reduced largely.
My GPU is GTX1080 8GB.

(Unfortunately, this computation is slower than numpy. We need more effort to optimize einsum.)

**Comment by fukatani** [comment]
Another option for introduce einsum optimization is use `numpy.einsum_path` directly.

To avoid numpy version dependence, I implemented einsum optimization for cupy.
(ex. numpy.1.9 doesn't supports einsum optimization.)

If PFN member recommend to use `numpy.einsum`, I will change design in this PR.

File: cupy/linalg/einsum.py [hunk 1]
```diff
@@ -305,45 +438,213 @@ def einsum(*operands):
...
+            sort_result = [(dim_dict[ind], ind) for ind in out_inds]
+            idx_result = ''.join([x[1] for x in sorted(sort_result)])
+
+        input_list.append(idx_result)
+        einsum_str = ','.join(tmp_inputs) + '->' + idx_result
+
+        contraction = (contract_inds, idx_removed, einsum_str, input_list[:])
+        contraction_list.append(contraction)
+
+    return operands, contraction_list
```
Comment (unnonouno) [inline cupy/linalg/einsum.py]: It looks that the function always behaves like `einsum_call = True`, right? So it is not compatible with the original one. I think you just want to optimize `einsum`, and it is not for users. It is better to add a comment about it in the docstring.

**Comment by fukatani** [comment]
>　Sorry for late reply. I tried to review the PR before, but it is too large for me.... I feel this PR contains some fixes, implementation of einsum_path, refactoring, and applying einsum_path to einsum. If possible can you split the PR?

Don't worry and sorry for too big PR.
Once pending some refactoring, I think there are two options for merging this PR.

1) Implement `cupy.einsum_path` by one PR, and applying it by another PR.
2) Use `numpy.einsum_path` (abolish cupy.einsum_path) and applying it in one PR.

pros. of 2)
We only need to merge one PR, so merging can be more quickly.
And if `numpy.einsum_path` will be improved, cupy einsum can benefit from it without additional commit.

cons. of 2)
`cupy.einsum` optiming functionality depends on `numpy` version.
Some old numpy version user can't use einsum  optimization.
Anf if `numpy.einsum_path` changes API, we have to handle all version API.

Do you have any opinion?

**Comment by fukatani** [comment]
Not only calculation time, the memory efficiency is greatly improved.
Since users can use more large size array in einsum, I think this PR is important.

**Comment by unnonouno** [comment]
I see. I checked the documentation of `einsum_path`, and is changed in recent version of NumPy. IMO it is ok to implement `cupy.einsum_path`.

**Comment by fukatani** [comment]
So, had I better do these steps?
1) Supporting `einsum_path` for internal. but not used in cupy.
2) Apply `einsum_path` to `cupy.einsum`.

**Comment by unnonouno** [comment]
That sounds easier to review for me (maybe)!

**Comment by fukatani** [comment]
I created #1008, implement einsum_path only. I close this PR. Thanks!

**Pull request closed without merging**